pkg/kubelet/cm/dra/plugin/registration_test.go

```diff
@@ -226,10 +266,87 @@ func TestRegistrationHandler(t *testing.T) {
 				// Nop.
 				handler.DeRegisterPlugin(test.pluginName, test.endpoint)
 
-				requireNoSlices()
+				requireNoSlices(t, tCtx, client)
 			})
 			assert.Equal(t, test.endpoint, plugin.endpoint, "plugin endpoint")
 			assert.Equal(t, test.chosenService, plugin.chosenService, "chosen service")
 		})
 	}
 }
+
+func TestConnectionHandling(t *testing.T) {
+	for description, test := range map[string]struct {
+		delay               time.Duration
+		requireSliceRemoval bool
+	}{
+		"wipe-slices-on-disconnect": {
+			delay:               time.Second, // very short wiping delay for testing
+			requireSliceRemoval: true,
+		},
+		"no-wipe-slices-on-reconnect": {
+			delay:               time.Minute, // long delay to avoid wiping
+			requireSliceRemoval: false,
+		},
+	} {
+		t.Run(description, func(t *testing.T) {
+			tCtx := ktesting.Init(t)
+
+			service := drapb.DRAPluginService
+			pluginName := "test-plugin"
+			sliceName := "test-slice"
+
+			slice := getSlice(sliceName)
+			client := getFakeClient(t, nodeName, pluginName, slice)
+
+			// The handler wipes all slices at startup.
+			handler := newRegistrationHandler(tCtx, client, getFakeNode, test.delay)
+			tCtx.Cleanup(handler.Stop)
+
+			// Run GRPC service
+			endpoint := path.Join(t.TempDir(), "dra-plugin-test.sock")
+			teardown, err := setupFakeGRPCServer(service, endpoint)
+			require.NoError(t, err)
+			defer teardown()
+
+			requireNoSlices(t, tCtx, client)
+
+			err = handler.RegisterPlugin(pluginName, endpoint, []string{service}, nil)
+			require.NoError(t, err)
+
+			err = WaitForConnection(tCtx, pluginName, endpoint, ConnectionPollInterval, ConnectionTimeout)
+			require.NoError(t, err, "failed to connect to plugin %s at %s", pluginName, endpoint)
+
+			plugin := draPlugins.get(pluginName)
+			assert.NotNil(t, plugin, "plugin should present in the plugin store")
+			assert.NotNil(t, plugin.conn, "plugin should be connected")
+
+			// Create the slice as if the plugin had done that while it runs.
+			_, err = client.ResourceV1beta1().ResourceSlices().Create(tCtx, slice, metav1.CreateOptions{})
+			require.NoError(t, err, "recreate slice")
+
+			// Simulate connection drop
+			require.NoError(t, plugin.conn.Close())
+
+			if test.requireSliceRemoval {
+				// Slice should be removed
+				requireNoSlices(t, tCtx, client)
+			} else {
+				// Simulate reconnect
+				plugin.closeConnection()
+				conn, err := plugin.getOrCreateGRPCConn()
```

Comment:
This is not a realistic test. What we need to test is that the gRPC connection monitoring detects a restarted gRPC server.

I've copied this test over into https://github.com/kubernetes/kubernetes/pull/132058 and made that change there.

Reply:
> This is not a realistic test.

This test is testing connection drop. What's unrealistic in it?